engine/access/rest/websockets/controller.go

```diff
@@ -4,100 +4,225 @@ import (
+	case <-ctx.Done():
+		// Context canceled, shut down gracefully
+		c.shutdownConnection()
+	}
+
+	// Ensure all goroutines finish execution.
+	wg.Wait()
+}
+
+// startProcess is a helper function to start a goroutine for a given process
+// and ensure it is tracked via a sync.WaitGroup.
+//
+// Parameters:
+// - wg: The wait group to track goroutines.
+// - ctx: The context for cancellation.
+// - process: The function to run in a new goroutine.
+//
+// No errors are expected during normal operation.
+func (c *Controller) startProcess(wg *sync.WaitGroup, ctx context.Context, process func(context.Context) error) {
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		err := process(ctx)
+		if err != nil {
+			// Check if shutdown has already been called, to avoid multiple shutdowns
+			if c.shutdown {
+				c.logger.Warn().Err(err).Msg("error detected after shutdown initiated, ignoring")
+				return
+			}
+
+			c.errorChannel <- err
+		}
+	}()
+}
+
+// configureConnection sets up the WebSocket connection with a read deadline
+// and a handler for receiving pong messages from the client.
+//
+// The function does the following:
+//  1. Sets an initial read deadline to ensure the server doesn't wait indefinitely
+//     for a pong message from the client. If no message is received within the
+//     specified `pongWait` duration, the connection will be closed.
+//  2. Establishes a Pong handler that resets the read deadline every time a pong
+//     message is received from the client, allowing the server to continue waiting
+//     for further pong messages within the new deadline.
+func (c *Controller) configureConnection() error {
+	// Set the initial read deadline for the first pong message
+	// The Pong handler itself only resets the read deadline after receiving a Pong.
+	// It doesn't set an initial deadline. The initial read deadline is crucial to prevent the server from waiting
+	// forever if the client doesn't send Pongs.
+	if err := c.conn.SetReadDeadline(time.Now().Add(PongWait)); err != nil {
+		return fmt.Errorf("failed to set the initial read deadline: %w", err)
+	}
+	// Establish a Pong handler which sets the handler for pong messages received from the peer.
+	c.conn.SetPongHandler(func(string) error {
+		return c.conn.SetReadDeadline(time.Now().Add(PongWait))
+	})
+
+	return nil
 }
 
 // writeMessagesToClient reads a messages from communication channel and passes them on to a client WebSocket connection.
 // The communication channel is filled by data providers. Besides, the response limit tracker is involved in
 // write message regulation
-func (c *Controller) writeMessagesToClient(ctx context.Context) {
-	//TODO: can it run forever? maybe we should cancel the ctx in the reader routine
+//
+// No errors are expected during normal operation.
+func (c *Controller) writeMessagesToClient(ctx context.Context) error {
 	for {
 		select {
 		case <-ctx.Done():
-			return
-		case msg := <-c.communicationChannel:
+			return ctx.Err()
```

Comment:
when the controller is updated to allow `c.communicationChannel` to be close when shutdown is triggered, you can remove this check, and instead rely on the channel closing to signal shutdown.

as it is now, if this returns here, the data providers may hang waiting to push onto the queue

Reply:
In updated `controller`, the idea of inclusion the `case <-ctx.Done(): return ctx.Err()` ensures alignment with `errgroup` behavior. When `errgroup.Wait()` detects an error in any goroutine, it cancels the parent `context`, signaling all child goroutines to stop. So goroutines must periodically check this `context` to exit early. This line allows the function to respect that cancellation.